### PR TITLE
[llvm][GlobalOpt] Remove empty atexit destructors/handlers

### PR DESCRIPTION
**Pull request by MaxEW707**
https://godbolt.org/z/frjhqMKqc for an example.

Removal of allocations due to empty `__cxa_atexit` destructor calls is done by the following globalopt pass.
This pass currently does not look for `atexit` handlers generated for platforms that do not use `__cxa_atexit`.
By default Win32 and AIX use `atexit`.

I don't see an easy way to only remove `atexit` calls that the compiler generated without looking at the generated mangled name of the atexit handler that is being registered.
However we can easily remove all `atexit` calls that register empty handlers since it is trivial to ensure the removed call still returns `0` which is the value for success.

### CODE REVIEW

File: llvm/test/tools/llvm-tli-checker/ps4-tli-check.yaml [hunk 1]
```diff
@@ -263,6 +263,10 @@ DynamicSymbols:
     Type:            STT_FUNC
     Section:         .text
     Binding:         STB_GLOBAL
+  - Name:            atexit
```
Comment (MaxEW707) [inline llvm/test/tools/llvm-tli-checker/ps4-tli-check.yaml:266]: @pogo59 @sks75220 Let me know for ps4/ps5if you prefer that this library function be set unavailable.

Comment (pogo59) [reply llvm/test/tools/llvm-tli-checker/ps4-tli-check.yaml:266]: `atexit` should be available for PS4/PS5, thanks!

File: llvm/lib/Transforms/IPO/GlobalOpt.cpp [hunk 2]
```diff
@@ -2321,14 +2322,16 @@ OptimizeGlobalAliases(Module &M,
...
+                  LibFunc Func) {
   // Hack to get a default TLI before we have actual Function.
   auto FuncIter = M.begin();
   if (FuncIter == M.end())
     return nullptr;
   auto *TLI = &GetTLI(*FuncIter);
 
-  LibFunc F = LibFunc_cxa_atexit;
+  LibFunc F = Func;
   if (!TLI->has(F))
```
Comment (nikic) [inline llvm/lib/Transforms/IPO/GlobalOpt.cpp:2335]:
```suggestion
  if (!TLI->has(Func))
```

and then declare `LibFunc F;` directly before the getLibFunc() call below, without initialization.

File: llvm/lib/Transforms/IPO/GlobalOpt.cpp [hunk 3]
```diff
@@ -2340,17 +2343,27 @@ FindCXAAtExit(Module &M, function_ref<TargetLibraryInfo &(Function &)> GetTLI) {
...
-static bool cxxDtorIsEmpty(const Function &Fn) {
+static Function *
+FindCXAAtExit(Module &M, function_ref<TargetLibraryInfo &(Function &)> GetTLI) {
+  return FindAtExitLibFunc(M, GetTLI, LibFunc_cxa_atexit);
+}
+
+static Function *
+FindAtExit(Module &M, function_ref<TargetLibraryInfo &(Function &)> GetTLI) {
+  return FindAtExitLibFunc(M, GetTLI, LibFunc_atexit);
+}
```
Comment (nikic) [inline llvm/lib/Transforms/IPO/GlobalOpt.cpp:2360]: These two wrapper functions look unnecessary, given how they are only used once.

File: llvm/test/Transforms/GlobalOpt/atexit-dtor.ll [hunk 4]
```diff
@@ -0,0 +1,46 @@
+; RUN: opt < %s -S -passes='cgscc(inline),function(early-cse),globalopt' | FileCheck %s
```
Comment (nikic) [inline llvm/test/Transforms/GlobalOpt/atexit-dtor.ll:1]: Change the test to only run globalopt, use update_test_checks.py.

File: llvm/test/Transforms/GlobalOpt/atexit-dtor.ll [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+%struct.A = type { i32 }
+
+$"??1A@@QEAA@XZ" = comdat any
+
+@"?g@@3UA@@A" = dso_local global %struct.A zeroinitializer, align 4
```
Comment (nikic) [inline llvm/test/Transforms/GlobalOpt/atexit-dtor.ll:7]: Use names without mangling please.

File: llvm/test/Transforms/GlobalOpt/atexit-dtor.ll (hunk 4 above)
Comment (MaxEW707) [reply llvm/test/Transforms/GlobalOpt/atexit-dtor.ll]: Done. Let me know if the changes are satisfactory.

Do note that I based my unit test off of, https://github.com/llvm/llvm-project/blob/main/llvm/test/Transforms/GlobalOpt/cxx-dtor.ll, which replicates the ir that the frontend generates for registering static objects with atexit.
This is why the `cgscc(inline)` pass was needed originally.

This pass assumes that other passes simplified the code enough as commented here, https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/IPO/GlobalOpt.cpp#L2351.

I am not familiar enough to know if unit testing such pass dependencies in unit tests is expected or if the unit test should be written as if other passes already simplified the ir.
I'll defer to your judgment there.

File: llvm/test/Transforms/GlobalOpt/atexit-dtor.ll [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+; Check that `atexit` is removed if the handler is empty.
+; Check that a removed `atexit` call returns `0` which is the value that denotes success.
+define dso_local noundef i32 @register_atexit_handler() {
+; CHECK-LABEL: define dso_local noundef i32 @register_atexit_handler() local_unnamed_addr {
+; CHECK-NEXT:    [[TMP1:%.*]] = alloca i32, align 4
+; CHECK-NEXT:    store i32 0, ptr [[TMP1]], align 4
+; CHECK-NEXT:    ret i32 0
+;
+  %1 = alloca i32, align 4
+  store i32 0, ptr %1, align 4
```
Comment (nikic) [inline llvm/test/Transforms/GlobalOpt/atexit-dtor.ll:22]: Can drop the alloca and store? They don't look related...

File: llvm/test/Transforms/GlobalOpt/atexit-dtor.ll [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  %2 = call i32 @atexit(ptr @"atexit_handler")
```
Comment (nikic) [inline llvm/test/Transforms/GlobalOpt/atexit-dtor.ll:23]:
```suggestion
  %2 = call i32 @atexit(ptr @atexit_handle")
```

File: llvm/test/Transforms/GlobalOpt/atexit-dtor.ll [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  ret i32 %2
```
Comment (nikic) [inline llvm/test/Transforms/GlobalOpt/atexit-dtor.ll:24]: Could you please add another call to atexit with a non-empty function as a negative test (i.e., to show that it will not be removed)?

**Review by nikic — APPROVED** [review]
LGTM
